File: app/src/main/java/com/mortarnav/MainActivity.java
package com.mortarnav;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;

import com.mortarnav.deps.WithActivityDependencies;
import com.mortarnav.mvp.home.HomeScreen;

import architect.Navigator;
import architect.NavigatorView;
import architect.commons.ActivityArchitector;
import architect.commons.Architected;
import architect.commons.transition.StandardTransition;
import architect.robot.DaggerService;
import autodagger.AutoComponent;
import autodagger.AutoInjector;
import butterknife.Bind;
import butterknife.ButterKnife;
import mortar.MortarScope;
import mortar.bundler.BundleServiceRunner;

/**
 * Root activity
 *
 * @author Lukasz Piliszczuk - lukasz.pili@gmail.com
 */
@AutoComponent(
        dependencies = App.class,
        superinterfaces = WithActivityDependencies.class
)
@AutoInjector
@DaggerScope(MainActivity.class)
public class MainActivity extends AppCompatActivity {

    private MortarScope scope;
    private Navigator navigator;

    @Bind(R.id.navigator_container)
    protected NavigatorView containerView;

    @Override
    public Object getSystemService(@NonNull String name) {
        if (scope != null && scope.hasService(name)) {
            return scope.getService(name);
        }

        return super.getSystemService(name);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        scope = ActivityArchitector.onCreateScope(this, savedInstanceState, new Architected() {
            @Override
            public Navigator createNavigator() {
                Navigator navigator = new Navigator(new Parceler());
                navigator.transitions().setDefault(new StandardTransition());
                return navigator;
            }

            @Override
            public void configureScope(MortarScope.Builder builder, MortarScope parentScope) {
                DaggerService.configureScope(builder, MainActivity.class, DaggerMainActivityComponent.builder()
                        .appComponent(parentScope.<AppComponent>getService(DaggerService.SERVICE_NAME))
                        .build());
            }
        });

        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        DaggerService.<MainActivityComponent>get(this).inject(this);

        // it is usually the best to create the navigator after everything else
        navigator = ActivityArchitector.onCreateNavigator(this, scope, savedInstanceState, containerView, new HomeScreen("Default home path"));
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        navigator.delegate().onNewIntent(intent);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        BundleServiceRunner.getBundleServiceRunner(scope).onSaveInstanceState(outState);
        navigator.delegate().onSaveInstanceState(outState);
    }

    @Override
    protected void onStart() {
        super.onStart();
        navigator.delegate().onStart();
    }

    @Override
    protected void onStop() {
        navigator.delegate().onStop();
        super.onStop();
    }

    @Override
    protected void onDestroy() {
        navigator.delegate().onDestroy();
        navigator = null;

        if (isFinishing() && scope != null) {
            scope.destroy();
            scope = null;
        }

        super.onDestroy();
    }

    @Override
    public void onBackPressed() {
        if (navigator.delegate().onBackPressed()) {
            return;
        }

        super.onBackPressed();
    }
}

File: kotlin_app/src/main/kotlin/architect/kotlinapp/MainActivity.kt
package architect.kotlinapp

import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.ViewGroup
import android.widget.FrameLayout
import architect.Architect
import architect.NavigationView
import architect.commons.ActivityArchitector
import architect.commons.Architected
import architect.commons.transition.StandardTransition
import architect.kotlinapp.mvp.home.screen.HomeScreen
import architect.kotlinapp.util.GlobalDependencies
import architect.kotlinapp.util.navigatorView
import architect.robot.dagger.DaggerScope
import architect.robot.dagger.DaggerService
import autodagger.AutoComponent
import autodagger.AutoInjector
import mortar.MortarScope
import mortar.bundler.BundleServiceRunner
import org.jetbrains.anko.UI

/**
 * @author Lukasz Piliszczuk - lukasz.pili@gmail.com
 */
@AutoComponent(dependencies = arrayOf(App::class), superinterfaces = arrayOf(GlobalDependencies::class))
@AutoInjector
@DaggerScope(MainActivity::class)
public class MainActivity : AppCompatActivity() {

    private var scope: MortarScope? = null
    private var architect: Architect? = null
    private lateinit var containerView: NavigationView

    override fun getSystemService(name: String?): Any? {
        return if (scope?.hasService(name) ?: false) scope?.getService(name) else super.getSystemService(name);
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        scope = ActivityArchitector.onCreateScope(this, savedInstanceState, object : Architected {
            override fun createNavigator(): Architect {
                val navigator = Architect(Parceler())
                navigator.transitions().setPushDefault(StandardTransition())
                return navigator
            }

            override fun configureScope(builder: MortarScope.Builder, parentScope: MortarScope) {
                DaggerService.configureScope(builder, MainActivity::class.java, DaggerMainActivityComponent.builder()
                        .appComponent(parentScope.getService<AppComponent>(DaggerService.SERVICE_NAME))
                        .build())
            }
        })

        setupView()

        DaggerService.get<MainActivityComponent>(this).inject(this)
        architect = ActivityArchitector.onCreateNavigator(this, scope, savedInstanceState, containerView, HomeScreen("test1", "test2"))
    }

    private fun setupView() {
        UI {
            containerView = navigatorView {
                layoutParams = FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
            }
        }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        architect?.delegate()?.onNewIntent(intent)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        BundleServiceRunner.getBundleServiceRunner(scope).onSaveInstanceState(outState)
        architect?.delegate()?.onSaveInstanceState(outState)
    }

    override fun onStart() {
        super.onStart()
        architect?.delegate()?.onStart()
    }

    override fun onStop() {
        architect?.delegate()?.onStop()
        super.onStop()
    }

    override fun onDestroy() {
        architect?.delegate()?.onDestroy()
        architect = null

        if (isFinishing) {
            scope?.destroy()
            scope = null
        }

        super.onDestroy()
    }

    override fun onBackPressed() {
        if (architect?.delegate()?.onBackPressed() ?: false) {
            return
        }

        super.onBackPressed()
    }
}
